### DISCUSSION TITLE
怎么和c语言一样在定义结构体的时候用宏控制？

### DISCUSSION THREAD

**Discussion by vkensou** [zigcc/forum#158] [新手提问] [answered]
比如c语言里这么定义：
```c
struct A
{
    int a;
#if _WIN64
    int b;
#endif
};
```
我想这么写是因为有时候结构体里有空洞，我想手动填充上。

**Comment by jiacai2050** [comment]
用一个编译时变量，控制选择哪个 struct，
```zig
const X = if (condition) struct {
    a: u64,
    b: bool,
} else struct {
    a: u64,
};
```

编译时变量可以参考：https://en.liujiacai.net/2023/06/29/embed-git-commit-in-zig/

**Reply by vkensou** [reply]
感觉有点重复

**Reply by jiacai2050** [reply]
这是我想到最简单的方案了......

**Comment by vkensou** [comment, marked as answer]
在ziggit上问到一个方法：
```
const A = struct {
    a: i32,
    b: switch (@import("builtin").os.tag) {
        .windows => i32,
        else => void,
    },
};
```
用`void`类型可以减少重复

**Reply by jiacai2050** [reply]
这个方法确实不错 👍

**Reply by hshq** [reply]
有点支离破碎的丑

**Reply by vkensou** [reply]
可以用函数封装下

**Reply by jiacai2050** [reply]
<img width="825" alt="image" src="https://github.com/user-attachments/assets/b6002de4-5de1-4f80-ab31-e8d264d6970f" />
感觉挺漂亮的，这是我之前代码的一个用法，说明 zig 代码的任意地方都可以是个表达式，很灵活。
